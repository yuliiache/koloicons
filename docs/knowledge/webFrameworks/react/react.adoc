== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







