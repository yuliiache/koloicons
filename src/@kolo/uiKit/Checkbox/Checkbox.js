import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import styles from './Checkbox.module.scss';

export const Checkbox = ({id, isChecked, onChange, isDisabled = false, hasError}) => {
  const errorMessage = 'Required';

  const checkboxClasses = classNames(styles.checkbox__indicator, {
    [styles['checkbox__indicator--error']]: hasError,
  });

  return (
      <label className={styles.checkbox}>
        <input
          className={styles.checkbox__input}
          id={id}
          type="checkbox"
          disabled={isDisabled}
          checked={isChecked}
          onChange={onChange}
        />
        <span className={checkboxClasses}></span>
        {hasError && <span className={styles.checkbox__error}>{errorMessage}</span>}
      </label>
  );
};


Checkbox.propTypes = {
  id: PropTypes.string.isRequired,
  isChecked: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
  isDisabled: PropTypes.bool,
  hasError: PropTypes.bool,
};

