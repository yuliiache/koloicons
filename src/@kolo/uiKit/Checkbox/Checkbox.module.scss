@import "../../design";

$checkmark-icon: url('./assets/check.svg');

.checkbox {
  font-size: .7rem;
  line-height: 1.43;
  color: $secFontColor;
  width: fit-content;
  position: relative;
  display: flex;
  align-items: center;
  margin: 10px;

  &__indicator {
    position: relative;
    box-sizing: border-box;
    justify-content: center;
    width: 20px;
    height: 20px;
    border-radius: 2px;
    margin-right: 8px;
    background: $WhiteBG none center no-repeat;
    background-size: contain;
    border: 1px solid $ActiveBorderColor;
    cursor: pointer;
    transition: background-color .2s, border-color .2s;
  }

  &__input {
    position: absolute;
    opacity: 0;
    cursor: pointer;
    height: 0;
    width: 0;

    &:focus + .checkbox__indicator {
      border-color: $BrandColor;
      box-shadow: 0 0 0 4px $FocusShadowInput;
    }

    &:focus + .checkbox__indicator--error {
      border-color: $errorColor;
      box-shadow: 0 0 0 4px rgba($errorColor,.2);
    }

    &:checked + .checkbox__indicator {
      background-color: $BrandColor;
      border-color: $BrandColor;
      background-image: $checkmark-icon;
    }

    &:disabled + .checkbox__indicator {
      cursor: auto;
      background-color: $DisabledColor;
      border-color: $ActiveBorderColor;

      &::after {
        content: "";
        position: absolute;
        top: 0;
        left: 0;
        right: 0;
        bottom: 0;
        background-image: $checkmark-icon;
        background-size: contain;
        background-repeat: no-repeat;
        background-position: center;
        filter: brightness(70%);
      }
    }
  }

  &__error {
    position: absolute;
    top: 100%;
    left: 0;
    font-size: 12px;
    line-height: 1.33;
    color: $errorColor;
  }
}
