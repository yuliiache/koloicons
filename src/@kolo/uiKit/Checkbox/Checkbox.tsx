import classNames from 'classnames';
import {FC} from 'react';

import styles from './Checkbox.module.scss';

interface Props {
  id: string;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  isChecked: boolean;
  hasError?: boolean;
  isDisabled?: boolean;
}

export const Checkbox: FC<Props> = ({id, onChange, isChecked, hasError = false, isDisabled = false}) => {
  const errorMessage = 'Required';

  const checkboxClasses = classNames(styles.checkbox__indicator, {
    [styles['checkbox__indicator--error']]: hasError,
  });

  return (
    <label className={styles.checkbox}>
      <input
        className={styles.checkbox__input}
        id={id}
        type="checkbox"
        disabled={isDisabled}
        checked={isChecked}
        onChange={onChange}
      />
      <span className={checkboxClasses}></span>
      {hasError && <span className={styles.checkbox__error}>{errorMessage}</span>}
    </label>
  );
};
