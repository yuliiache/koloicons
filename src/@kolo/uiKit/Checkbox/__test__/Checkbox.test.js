import React from 'react';
import { cleanup,fireEvent, render, screen } from '@testing-library/react';

import '@testing-library/jest-dom';

import { Checkbox} from "../Checkbox";

describe('Checkbox Component', () => {
  afterEach(() => {
    cleanup();
  }
  );

  it('basic render succeed', () => {
    const handleChange = jest.fn();
    const { asFragment } = render(<Checkbox id='testId' isChecked={false} onChange={handleChange} />);
    expect(screen.getByRole('checkbox')).toHaveAttribute('id');
    expect(asFragment()).toMatchSnapshot();
    }
  );

  it('handles isChecked state', () => {
    const handleChange = jest.fn();
    const { rerender, asFragment } = render(<Checkbox id='testId' isChecked={true} onChange={handleChange} />);
    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toBeChecked();
    expect(asFragment()).toMatchSnapshot();


      fireEvent.click(checkbox);
    expect(handleChange).toHaveBeenCalledTimes(1);

    rerender(<Checkbox id='testId' isChecked={true} onChange={handleChange} />);
    expect(checkbox).toBeChecked();
    expect(asFragment()).toMatchSnapshot();
    }
  );

  it('triggers onChange event', () => {
    const handleChange = jest.fn();
    const { asFragment } = render(<Checkbox id='testId' isChecked={false} onChange={handleChange} />);
    const checkbox = screen.getByRole('checkbox');
    fireEvent.click(checkbox);
    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(asFragment()).toMatchSnapshot();
    }
  );

  it('respects isDisabled prop', () => {
    const handleChange = jest.fn();
    const { asFragment } = render(<Checkbox id='testId' isChecked={true} isDisabled={true} onChange={handleChange} />);
    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toBeDisabled();
    expect(asFragment()).toMatchSnapshot();

    }
  );

  it('shows error message when hasError is true', () => {
    const handleChange = jest.fn();
    const { asFragment } = render(<Checkbox id='testId' isChecked={false} onChange={handleChange} hasError={true} />);
    expect(screen.getByText('Required')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
    }
  );
}
);
