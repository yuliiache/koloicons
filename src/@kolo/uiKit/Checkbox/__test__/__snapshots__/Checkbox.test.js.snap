// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Component basic render succeed 1`] = `
<DocumentFragment>
  <label
    class="checkbox"
  >
    <input
      class="checkbox__input"
      id="testId"
      type="checkbox"
    />
    <span
      class="checkbox__indicator"
    />
  </label>
</DocumentFragment>
`;

exports[`Checkbox Component handles isChecked state 1`] = `
<DocumentFragment>
  <label
    class="checkbox"
  >
    <input
      checked=""
      class="checkbox__input"
      id="testId"
      type="checkbox"
    />
    <span
      class="checkbox__indicator"
    />
  </label>
</DocumentFragment>
`;

exports[`Checkbox Component handles isChecked state 2`] = `
<DocumentFragment>
  <label
    class="checkbox"
  >
    <input
      checked=""
      class="checkbox__input"
      id="testId"
      type="checkbox"
    />
    <span
      class="checkbox__indicator"
    />
  </label>
</DocumentFragment>
`;

exports[`Checkbox Component respects isDisabled prop 1`] = `
<DocumentFragment>
  <label
    class="checkbox"
  >
    <input
      checked=""
      class="checkbox__input"
      disabled=""
      id="testId"
      type="checkbox"
    />
    <span
      class="checkbox__indicator"
    />
  </label>
</DocumentFragment>
`;

exports[`Checkbox Component shows error message when hasError is true 1`] = `
<DocumentFragment>
  <label
    class="checkbox"
  >
    <input
      class="checkbox__input"
      id="testId"
      type="checkbox"
    />
    <span
      class="checkbox__indicator checkbox__indicator--error"
    />
    <span
      class="checkbox__error"
    >
      Required
    </span>
  </label>
</DocumentFragment>
`;

exports[`Checkbox Component triggers onChange event 1`] = `
<DocumentFragment>
  <label
    class="checkbox"
  >
    <input
      class="checkbox__input"
      id="testId"
      type="checkbox"
    />
    <span
      class="checkbox__indicator"
    />
  </label>
</DocumentFragment>
`;
